### DISCUSSION TITLE
Why are are backends in Split spec k8s services

### DISCUSSION THREAD

**Discussion by shashankram** [servicemeshinterface/smi-spec#164] [General]
Per my understanding, K8s services are abstractions for clients to communicate with a service without having to worry about the backend endpoints for the service. A client would use the service FQDN and resolve to a cluster IP, which K8s would use to load-balance traffic to backends running the service over their endpoints. This allows clients to continue using the service FQDN even if their backends change.

With SMI Split, the user needs to define additional K8s services for the backends. Had the backends been exposed in the form of selectors similar to Istio, users could continue to use a Service as an abstraction for the client apps. If users need to use Split, they need to define additional services which no longer provide the abstraction a K8s Service resource provides.

What was the reason to do this? Are users okay to define additional services just for Split? Another benefit with using the Service abstraction and not enforcing the backends in the Split to be services is that as the backends are versioned out, clients can continue to use the abstract Service FQDN without having to maintain version specific Service resources.

**Comment by grampelberg** [comment]
> If users need to use Split, they need to define additional services which no longer provide the abstraction a K8s Service resource provides.

I don't understand this. They're just k8s services, so wouldn't they provide the same abstraction?

**Reply by shashankram** [reply]
Lets take the example where a user has a service `website` running in the `default` namespace. With Kubernetes, a user would create a `website` service and deployments with app selector set to `website`.  Clients accessing the `website` service will use the service FQDN of `website.default.svc.cluster.local`, which would resolve to a cluster IP. Requests from clients to this FQDN will be load balanced to the backends hosting the service. So there is a single `website` service abstraction for multiple backends.

With SMI, the user would need to create a k8s service resource per backend, and also possibly a service resource just for the domain to be resolvable. So if the user wants to Split traffic to `website.default.svc.cluster.local` between `website-v1` and `website-v2`, corresponding service resources for the backend need to be created. This means a service resource no longer provides abstraction to its backends.

**Reply by grampelberg** [reply]
Isn't it true that there's simply more configuration? The user can set the selector for `website` to whatever they want (v1, v2, the entire pool). At some point, the selectors need to get configured, the question is what object owns that configuration.

FWIW, I think the other side of this is valid as well. With Istio, there is no service abstraction there for the new backends. If you wanted to hit those backends explicitly for testing, you wouldn't be able to do `website-v2` (especially from a non-meshed resource).

**Reply by grampelberg** [reply]
This is an implementation argument, but we get a lot from using services. Off the top of my head:

- Readiness checks automatically are handled and add/remove endpoints based off service readiness.
- Endpoints get service topologies automatically.
- It works outside of the mesh, allowing ways to get out of an incident.

**Reply by shashankram** [reply]
Thanks @grampelberg for your response. My primary concern is around users having to deploy multiple service resources for the backends which IMO breaks the abstraction a K8s service provides. Yes the selector would need to be configured, but the abstraction of a service remains the same, allowing clients to use 1 FQDN throughout the lifetime of that service.

With SMI, the root level service would need to continue to live in K8s and client code even if they are versioned out with newer versions of the service.

**Reply by grampelberg** [reply]
> allowing clients to use 1 FQDN throughout the lifetime of that service.

I don't understand, why would you need to use a new FQDN for the clients?

**Reply by shashankram** [reply]
Clients would communicate with the service using their FQDN `service.namespace.svc.cluster.local`.

**Reply by grampelberg** [reply]
Right, so it doesn't change either for traffic split or istio.

**Reply by shashankram** [reply]
Yes the top level service in Split is the service that needs to exist at all times. My point is that the backends need not be services. Making the backends services means that a single service can no longer provide abstraction to multiple backends hosting the service.

**Reply by grampelberg** [reply]
Is there a way to make a single k8s service provide abstraction to multiple backends?

**Reply by shashankram** [reply]
This should be possible using label selectors. Consider the following example.
Service `website` with backends `website-v1` and `website-v2`.

A single service resource with the name `website` would exist. Deployments with label selectors `app: website`, `version: <website-version>` would exist, essentially tying the backends (pods) to the service.

The traffic Split definition could possibly look as follows:
```
apiVersion: split.smi-spec.io/v1alpha3
kind: TrafficSplit
metadata:
  name: website-canary
  namespace: default
spec:
  service: website
  backends:
  - version: website-v1
    weight: 90
  - version: website-v2
    weight: 10
```

With something like the above, the service `website` is the only service resource that exists, the backends are abstracted. The `website` service resource can live even as the backends are versioned out. Clients accessing the website service are only using `website.default.svc.cluster.local` at all times, and never directly reference the backends. This provides the same degree of abstraction K8s services provide for load balancing traffic to the backends of `website` service using their cluster IP, the differenence in the service mesh being that the load balancing is determined by the SMI Split spec, in this case the weights assigned to the versions.

**Reply by shashankram** [reply]
@grampelberg does the above make sense to you?

**Reply by grampelberg** [reply]
I'm having a tough time understanding the benefits, namely your point:

> Clients accessing the website service are only using website.default.svc.cluster.local at all times, and never directly reference the backends. 

This is how it works today with the existing implementation.

> This provides the same degree of abstraction K8s services provide for load balancing traffic to the backends of website service using their cluster IP,

This is not actually true, as `ClusterIP` will forward to *all* backends - both v1 and v2, which could be disastrous for some users.

There *might* be something we could do around service topologies to make this work, but I'm concerned that we're overloading concepts at that point.

**Reply by shashankram** [reply]
> I'm having a tough time understanding the benefits, namely your point:
> 
> > Clients accessing the website service are only using website.default.svc.cluster.local at all times, and never directly reference the backends.
> 
> This is how it works today with the existing implementation.

Yes this part is well and good. I am complaining about users having to define a service per backend which breaks the whole point of defining a service as an abstraction for its backends.

> 
> > This provides the same degree of abstraction K8s services provide for load balancing traffic to the backends of website service using their cluster IP,
> 
> This is not actually true, as `ClusterIP` will forward to _all_ backends - both v1 and v2, which could be disastrous for some users.
> 

You misunderstood my comment. I am not suggesting the service mesh do this, but suggesting that the service mesh refer to the TrafficSplit spec to load balance traffic. The difference in my example spec above is that the backend are not services, just versions.

> There _might_ be something we could do around service topologies to make this work, but I'm concerned that we're overloading concepts at that point.

I am trying to convey that making the backends refer to service resources via their names breaks the service abstraction kubernetes provides for its backends. I am suggesting we use `version` or some other selector for `backends` instead of `service`.

**Reply by shashankram** [reply]
@grampelberg I still see a value in using selectors for backends. Is this something that can be supported in addition to having the backends be services?

**Reply by grampelberg** [reply]
Totally up for the discussion! I'll be honest, I don't see the value and the complexity it adds to implementations is pretty steep. Especially when you consider that it'll then miss out on service topologies and readiness checks.

We could make the field an object reference instead of an actual service name. That'd allow you to implement your own selector CRD and make it optional to support for implementations.

**Reply by shashankram** [reply]
Creating a service per backend is confusing for users. Providing additional flexibility would be great. Interested in learning more about your proposal above. How would a sample spec look like with object references?

**Reply by grampelberg** [reply]
```yaml
apiVersion: split.smi-spec.io/v1alpha3
kind: TrafficSplit
metadata:
  name: website-canary
  namespace: default
spec:
  service: website
  backends:
  - resource:
       kind: Selector
       name: website-v1
     weight: 90
  - resource:
       kind: Selector
       name: website-v2
     weight: 10
---
kind: Selector
metadata:
  name: website-v1
  namespace: default
spec:
  selector:
    version: website-v1
    name: website
---
kind: Selector
metadata:
  name: website-v2
  namespace: default
spec:
  selector:
    version: website-v2
    name: website
```

This would obviously only be supported by your implementation of SMI, as anything can be an object reference (and it ends up being a subset of the functionality that you'd get by using services), but it'd get you what you want.
